Convert export-w3c-test-changes to use webkitscmpy/webkitbugspy

briannafan opened the pull request:
#### 56c33b4b15944cd3010ed1da9e6c3f2dbfdcce7a
<pre>
Convert export-w3c-test-changes to use webkitscmpy/webkitbugspy
<a href="https://rdar.apple.com/110276747">rdar://110276747</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=257728">https://bugs.webkit.org/show_bug.cgi?id=257728</a>

Reviewed by Sam Sneddon.

Move all git operations to the webkitscmpy.local.Git class instead of webkitpy.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
    - Add mocks for `git apply&apos; and `git commit -m`.

* Tools/Scripts/webkitpy/tool/commands/upload_unittest.py:
   - Add new logging to tests

* Tools/Scripts/webkitpy/w3c/test_exporter.py:
(WebPlatformTestExporter.__init__):
(WebPlatformTestExporter.username):
(WebPlatformTestExporter._wpt_fork_push_url):
(WebPlatformTestExporter._init_wpt_remote):
(WebPlatformTestExporter):
(WebPlatformTestExporter._run_wpt_git):
(WebPlatformTestExporter.has_wpt_changes):
(WebPlatformTestExporter._ensure_wpt_repository):
(WebPlatformTestExporter._ensure_new_branch_name):
(WebPlatformTestExporter.clean):
(WebPlatformTestExporter.create_branch_with_patch):
(WebPlatformTestExporter.set_up_wpt_fork):
(WebPlatformTestExporter.push_to_wpt_fork):
(WebPlatformTestExporter.make_pull_request):
(WebPlatformTestExporter.create_wpt_pull_request):
(WebPlatformTestExporter.delete_local_branch):
(WebPlatformTestExporter.do_export):
(parse_args): Remove unused arguments.

(main):
(WebPlatformTestExporter.token): Deleted.
(WebPlatformTestExporter._github): Deleted.
(WebPlatformTestExporter._git): Deleted.
(WebPlatformTestExporter._prompt_for_token): Deleted.
(WebPlatformTestExporter._prompt_for_username): Deleted.
(WebPlatformTestExporter._ensure_username_and_token): Deleted.
(WebPlatformTestExporter._validate_and_save_token): Deleted.
(WebPlatformTestExporter._fetch_wpt_repository): Deleted.
(WebPlatformTestExporter.create_upload_remote_if_needed): Deleted.

* Tools/Scripts/webkitpy/w3c/test_exporter_unittest.py:
(TestExporterTest):
(TestExporterTest.MockGit):
(TestExporterTest.MockGit.__init__):
(TestExporterTest.test_export):
(TestExporterTest.test_export_with_specific_branch):
(TestExporterTest.test_export_no_clean):
(TestExporterTest.test_export_interactive_mode):
(TestExporterTest.test_export_invalid_token):
(TestExporterTest.test_export_wrong_token):
(TestExporterTest.test_has_wpt_changes):
(TestExporterTest.test_has_no_wpt_changes_for_no_diff):
(TestExporterTest.test_ignore_changes_to_expected_file):
(test_ignore_changes_to_expected_mismatch_file):
(test_ignore_changes_to_w3c_import_log):
(TestExporterTest.MockGit.clone): Deleted.
(TestExporterTest.MockGit.fetch): Deleted.
(TestExporterTest.MockGit.checkout): Deleted.
(TestExporterTest.MockGit.reset_hard): Deleted.
(TestExporterTest.MockGit.push): Deleted.
(TestExporterTest.MockGit.delete_branch): Deleted.
(TestExporterTest.MockGit.checkout_new_branch): Deleted.
(TestExporterTest.MockGit.apply_mail_patch): Deleted.
(TestExporterTest.MockGit.commit): Deleted.
(TestExporterTest.MockGit.remote): Deleted.
(TestExporterTest.MockGit.local_config): Deleted.
(TestExporterTest.MockGit.branch_ref_exists): Deleted.

Canonical link: <a href="https://commits.webkit.org/296936@main">https://commits.webkit.org/296936@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4d84d1c79b10cc9b2f5a147738c57de58d5dfe1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110042 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20133 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116064 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112005 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30379 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38288 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/83661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112990 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/24224 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/99099 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/64104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/109474 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/23593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/17241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/59860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/93597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/17290 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118855 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37082 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/109538 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37454 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/95367 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/92458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37443 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/15182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17754 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36976 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36638 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39978 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/38347 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
```diff
@@ -578,6 +578,15 @@ def editor_generator(*args, **kwargs):
+                cwd=self.path,
+                generator=lambda *args, **kwargs: self.apply(),
+            ),
+            mocks.Subprocess.Route(
```
While I'd much prefer we use this indentation everywhere, it doesn't match everything above it (boo!).

briannafan replied:
Oops! thanks

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -34,6 +34,9 @@
 from webkitscmpy import local, log, remote
 
+from webkitpy.w3c.test_exporter import WebPlatformTestExporter, parse_args, configure_logging
+from webkitpy.common.host import Host
```
I remember a while ago @JonWBedard having opinions about the generality of `git-webkit`; this is kinda inherently tied to the WebKit repo — we currently don't have any dependency on `webkitpy` from `webkitscmpy`.

We almost certainly want to make any import fallible, because otherwise it'll cease being possible to run the `webkitscmpy` tests in isolation, and potentially use it outside of WebKit.

briannafan replied:
That's a good point! We can discuss this and I'll keep it in mind with the next PR for `git-webkit` integration. This was mainly some quick prototyping on my part.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -740,6 +743,14 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
             if args.open:
                 Terminal.open_url(pr.url)
 
+        log.info('Checking for WPT changes to export...\n')
```
Why the trailing new line?

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
```diff
@@ -740,6 +743,14 @@ def create_pull_request(cls, repository, args, branch_point, callback=None, unbl
 
+        log.info('Checking for WPT changes to export...\n')
+        options = parse_args(['--git-commit', commits[0].hash, '--create-pr'])
+        configure_logging()
```
Do we actually want to use this? I'd expect that we'd want to use our own config here, to try and be close to what `webkitscmpy` does?

briannafan replied:
Yeah, I was just speed prototyping with all the changes in `pull_request.py`. But your comments will be helpful for the next PR!

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -54,28 +52,30 @@
+                issue = next((issue for issue in commit.issues if isinstance(issue.tracker, bugzilla.Tracker)), None)
+                if not issue:
+                    _log.error('Unable to find associated bug.')
+                    sys.exit()
```
I'd assume we want to exit with something non-zero?

That said, I do wonder if we want to be calling `sys.exit` from here at all — rather than raising some exception, then catching it within `webkitscmpy.program` to then exit with.

`sys.exit` isn't ideal if this ends up being used elsewhere.

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -89,6 +89,11 @@ def __init__(self, host, options, gitClass=Git, bugzillaClass=Bugzilla, WPTGitHu
 
+        if not self._filesystem.exists(self._options.repository_directory):
+            run([local.Git.executable(), 'clone', f'{WPT_GH_URL}.git', self._options.repository_directory], cwd=os.path.dirname(os.getcwd()))
+        self._wpt_repo = local.Git(self._options.repository_directory)
```
This means we more eagerly create the repo than we previously did, right? Because this used to happen in the probably now dead `_ensure_wpt_repository`.

This means creating an exporter just to call `wpt_exporter.has_wpt_changes()` already clones the repo if it doesn't exist, which doesn't seem ideal.

briannafan replied:
Good point. I'll move this logic so it's only called when we have WPT changes and want to export.

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -164,7 +151,22 @@ def _wpt_patch(self):
+    def _init_wpt_remote(self):
+        source_remote = self._wpt_repo.default_remote
+        if not self._wpt_repo.config().get('remote.{}.url'.format(source_remote)):
+            _log.error("'{}' is not a remote in this repository\n".format(source_remote))
```
Why the trailing new line?

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -164,7 +151,22 @@ def _wpt_patch(self):
+            return
+        remote = self._wpt_repo.remote(name=source_remote)
+        if not remote:
+            _log.error("'{}' doesn't have a recognized remote\n".format(self._wpt_repo.root_path))
```
Why the trailing new line?

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +214,134 @@ def write_git_patch_file(self):
+            ['remote', 'add', self._wpt_fork_remote, self._wpt_fork_push_url]
+        ).returncode not in [0, 3]:
+            _log.error("Failed to add '{}' remote\n".format(self._wpt_fork_remote))
+            return
```
Can we not just check that `f"remote.{self._wpt_fork_remote}.url"` exists in the config, as we do above? I don't know how `Git.source_remotes(…)` works though.

briannafan replied:
Yeah! That gets rid of a lot of extra WebKit specific logic we don't need

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +214,134 @@ def write_git_patch_file(self):
         description = f'WebKit export from bug: [{title}]({self._bug.link})'
-        pr_number = self.create_wpt_pull_request(self._wpt_fork_remote + ':' + self._public_branch_name, self._commit_message, description)
-        if pr_number:
+        pr = self.create_wpt_pull_request(self._wpt_fork_remote + ':' + self._public_branch_name, self._commit_message, description)  # TODO
```
What's the TODO here?

briannafan replied:
To remove the TODO 😛

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +214,134 @@ def write_git_patch_file(self):
-                self._github.add_label(pr_number, WEBKIT_EXPORT_PR_LABEL)
+                pr_issue = pr._metadata['issue']
+                labels = pr_issue.labels
+                labels.append(WEBKIT_EXPORT_PR_LABEL)
```
None of this needs to be in the `try .. except` block, only the last line when we try to actually set the labels.

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +214,134 @@ def write_git_patch_file(self):
-        pr_number = self.create_wpt_pull_request(self._wpt_fork_remote + ':' + self._public_branch_name, self._commit_message, description)
-        if pr_number:
+        pr = self.create_wpt_pull_request(self._wpt_fork_remote + ':' + self._public_branch_name, self._commit_message, description)  # TODO
+        if pr and pr._metadata and pr._metadata.get('issue'):
```
Presumably we should just have some `if pr` and handle the case where the PR has failed to be created, returning early, rather than having `if pr and ..` multiple times here?

briannafan replied:
We would have to check for the metadata and issue being populated correctly anyways - not sure if returning early would save much. If the pr doesn't exist, we'd return None immediately anyways.

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +214,134 @@ def write_git_patch_file(self):
+            head=self._wpt_repo.branch,
+        )
+        if not pr:
+            sys.stderr.write("Failed to create pull-request for '{}'\n".format(self._wpt_repo.branch))
```
Why do we sometimes just write to stderr, and sometimes rely on logging output? I'm not sure I understand the basis of this division.

briannafan replied:
I lazily copied this over from `webkitscmpy` - I'll go through and fix them

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +214,134 @@ def write_git_patch_file(self):
+            sys.stderr.write("Failed to create pull-request for '{}'\n".format(self._wpt_repo.branch))
+            return None
+
+        print("Created '{}'!".format(pr))
```
Why not an f-string?

gsnedders reviewed (changes requested):
This probably isn't everything, but probably is enough to start some discussion? (Sorry! A lot of these are questions, though, rather than definite don't-do-this!)

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -164,7 +141,22 @@ def _wpt_patch(self):
 
+    def _init_wpt_remote(self):
+        source_remote = self._wpt_repo.default_remote
+        if not self._wpt_repo.config().get('remote.{}.url'.format(source_remote)):
```
Any reason to not use an f-string here?

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +204,147 @@ def write_git_patch_file(self):
-        _log.info('Fetching web-platform-tests repository')
-        self._git.fetch()
+        if not self._filesystem.exists(self._options.repository_directory):
+            run([local.Git.executable(), 'clone', f'{WPT_GH_URL}.git', self._options.repository_directory], cwd=os.path.dirname(os.getcwd()))
```
`cwd=os.path.dirname(os.getcwd())` is completely redundant; you're just passing it with what it already is.

We could almost it not rely on the CWD at all by making `repository_directory` absolute.

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +204,147 @@ def write_git_patch_file(self):
         _log.info('Cleaning web-platform-tests master branch')
-        self._git.checkout('master')
-        self._git.reset_hard('origin/master')
+        self._run_wpt_git(['checkout', self._wpt_repo.default_branch])
```
Do we really not have any API for this?

briannafan replied:
We do, but there's a lot of WebKit specific logic in it that I wasn't sure we wanted to introduce here.

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +204,147 @@ def write_git_patch_file(self):
+            return
+
+        if self._run_wpt_git(['remote', 'show', self._wpt_fork_remote], capture_output=True).returncode:
+            # FIXME: If a fork doesn't exist, we should create one for the user.
```
Link to the Bugzilla issue.

gsnedders commented on Tools/Scripts/webkitpy/w3c/test_exporter.py:
```diff
@@ -212,173 +204,147 @@ def write_git_patch_file(self):
-        while self._git.branch_ref_exists(branch_name):
+        branches_for = self._wpt_repo.branches_for()
+        while branch_name in branches_for:
+            # FIXME: If the branch exists, we should give the option to overwrite or rebase.
```
Do we have a Bugzilla link for this? Because including that helps with tracking.

briannafan replied:
Created one!

gsnedders reviewed (approved):
These are all minor nits; I'll let you decide what to do (and/or whether you want to seek further review while I'm out of any changes, though I'd expect any changes to be minor).

briannafan committed:
Convert export-w3c-test-changes to use webkitscmpy/webkitbugspy
rdar://110276747
https://bugs.webkit.org/show_bug.cgi?id=257728

Reviewed by Sam Sneddon.

Move all git operations to the webkitscmpy.local.Git class instead of webkitpy.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/local/git.py:
    - Add mocks for `git apply' and `git commit -m`.

* Tools/Scripts/webkitpy/tool/commands/upload_unittest.py:
   - Add new logging to tests

* Tools/Scripts/webkitpy/w3c/test_exporter.py:
(WebPlatformTestExporter.__init__):
(WebPlatformTestExporter.username):
(WebPlatformTestExporter._wpt_fork_push_url):
(WebPlatformTestExporter._init_wpt_remote):
(WebPlatformTestExporter):
(WebPlatformTestExporter._run_wpt_git):
(WebPlatformTestExporter.has_wpt_changes):
(WebPlatformTestExporter._ensure_wpt_repository):
(WebPlatformTestExporter._ensure_new_branch_name):
(WebPlatformTestExporter.clean):
(WebPlatformTestExporter.create_branch_with_patch):
(WebPlatformTestExporter.set_up_wpt_fork):
(WebPlatformTestExporter.push_to_wpt_fork):
(WebPlatformTestExporter.make_pull_request):
(WebPlatformTestExporter.create_wpt_pull_request):
(WebPlatformTestExporter.delete_local_branch):
(WebPlatformTestExporter.do_export):
(parse_args): Remove unused arguments.

(main):
(WebPlatformTestExporter.token): Deleted.
(WebPlatformTestExporter._github): Deleted.
(WebPlatformTestExporter._git): Deleted.
(WebPlatformTestExporter._prompt_for_token): Deleted.
(WebPlatformTestExporter._prompt_for_username): Deleted.
(WebPlatformTestExporter._ensure_username_and_token): Deleted.
(WebPlatformTestExporter._validate_and_save_token): Deleted.
(WebPlatformTestExporter._fetch_wpt_repository): Deleted.
(WebPlatformTestExporter.create_upload_remote_if_needed): Deleted.

* Tools/Scripts/webkitpy/w3c/test_exporter_unittest.py:
(TestExporterTest):
(TestExporterTest.MockGit):
(TestExporterTest.MockGit.__init__):
(TestExporterTest.test_export):
(TestExporterTest.test_export_with_specific_branch):
(TestExporterTest.test_export_no_clean):
(TestExporterTest.test_export_interactive_mode):
(TestExporterTest.test_export_invalid_token):
(TestExporterTest.test_export_wrong_token):
(TestExporterTest.test_has_wpt_changes):
(TestExporterTest.test_has_no_wpt_changes_for_no_diff):
(TestExporterTest.test_ignore_changes_to_expected_file):
(test_ignore_changes_to_expected_mismatch_file):
(test_ignore_changes_to_w3c_import_log):
(TestExporterTest.MockGit.clone): Deleted.
(TestExporterTest.MockGit.fetch): Deleted.
(TestExporterTest.MockGit.checkout): Deleted.
(TestExporterTest.MockGit.reset_hard): Deleted.
(TestExporterTest.MockGit.push): Deleted.
(TestExporterTest.MockGit.delete_branch): Deleted.
(TestExporterTest.MockGit.checkout_new_branch): Deleted.
(TestExporterTest.MockGit.apply_mail_patch): Deleted.
(TestExporterTest.MockGit.commit): Deleted.
(TestExporterTest.MockGit.remote): Deleted.
(TestExporterTest.MockGit.local_config): Deleted.
(TestExporterTest.MockGit.branch_ref_exists): Deleted.

Canonical link: https://commits.webkit.org/296936@main
